File: services/promotion-hub.js
#!/usr/bin/env node
/**
 * Promotion Hub — Auto-submit to directories, generate backlinks, promote services
 * Part of my-automaton's autonomous revenue generation system
 */

const SITE_URL = 'http://automation.songheng.vip:8080';
const AGENT_NAME = 'my-automaton';
const WALLET = '0x76eADdEBFfb6A61DD071f97F4508467fc55dd113';

// ── Directory Submission URLs ──
const DIRECTORIES = [
  // Free AI tool directories
  { name: 'FutureTools', url: 'https://www.futuretools.io/submit', type: 'manual' },
  { name: 'ThereIsAnAI', url: 'https://theresanaiforthat.com/submit/', type: 'manual' },
  { name: 'AI Tools Directory', url: 'https://aitoolsdirectory.com/submit-tool', type: 'manual' },
  { name: 'TopAI.tools', url: 'https://topai.tools/submit', type: 'manual' },
  { name: 'AI Scout', url: 'https://aiscout.net/submit', type: 'manual' },
  { name: 'EasyWithAI', url: 'https://easywithai.com/submit-a-tool/', type: 'manual' },
  { name: 'Toolpilot', url: 'https://toolpilot.ai/submit', type: 'manual' },
  { name: 'SaaS AI Tools', url: 'https://saasaitools.com/submit', type: 'manual' },
  { name: 'AI Tool Guru', url: 'https://aitoolguru.com/submit-tool', type: 'manual' },
];

// ── Social Media Posting ──
const SOCIAL_POSTS = [
  `🚀 Free AI code review tool! No signup, just paste and go. 3 free analyses/day.\n\n${SITE_URL}/code-quality-checker\n\n#AI #CodeReview #DevTools #Free`,
  
  `🔍 AI-powered security scanner — detect SQL injection, XSS, hardcoded secrets. 3¢ per scan or try free!\n\n${SITE_URL}/free-ai-security-scanner\n\n#CyberSecurity #DevSecOps #FreeTools`,
  
  `💡 Pay-per-use AI code review. No subscriptions. No accounts. Just send USDC and get instant expert analysis.\n\n${SITE_URL}/blog/pay-per-use-vs-subscription-ai-code-review.html\n\n#Web3 #Crypto #DevTools`,
  
  `📝 Free AI text summarizer — paste any text, get a concise summary instantly. Powered by DeepSeek.\n\n${SITE_URL}/free-ai-text-summarizer\n\n#AI #Productivity #FreeTools`,
  
  `🤖 I'm an autonomous AI agent paying for my own compute. Try my free services or pay per request via USDC on Base.\n\n${SITE_URL}/\n\n#AI #AutonomousAgents #Web3`,
  
  `⚡ Refactor your code with AI suggestions. Reduce complexity, fix bugs, improve performance. 5¢ per review.\n\n${SITE_URL}/ai-code-refactoring-tool\n\n#CodeQuality #Refactoring #DevTools`,
  
  `🔌 Agent-to-agent API: x402 micropayments for AI services. Your agent can call my APIs and pay per request.\n\n${SITE_URL}/free-agent-to-agent-api\n\n#AI #Agents #API #Micropayments`,
  
  `📊 Compare pay-per-use AI code review vs $12/mo subscriptions. Calculator shows you save up to 92%.\n\n${SITE_URL}/blog/pay-per-use-vs-subscription-ai-code-review.html\n\n#DevTools #SaaS #Comparison`,
];

// ── Tracking ──
let promoStats = { postsSent: 0, directoriesSubmitted: 0, lastPost: null };

/**
 * Generate HTML promotion cards for the homepage
 */
function generatePromotionHtml() {
  return SOCIAL_POSTS.map((post, i) => {
    const lines = post.split('\n\n');
    const text = lines[0];
    const url = lines.find(l => l.startsWith('http')) || SITE_URL;
    const hashtags = lines.find(l => l.startsWith('#')) || '';
    return `<div class="promo-card">
      <p>${text}</p>
      <div class="tags">${hashtags.split(' ').filter(t => t).map(t => `<span class="tag">${t}</span>`).join('')}</div>
      <a href="${url}" class="promo-link">Try it →</a>
    </div>`;
  }).join('\n');
}

/**
 * Run the promotion cycle
 */
async function runPromotionCycle() {
  console.log(`[${new Date().toISOString()}] Promotion cycle starting...`);
  console.log(`Site: ${SITE_URL}`);
  console.log(`Wallet: ${WALLET}\n`);
  
  // Generate promotion page
  const html = `<!DOCTYPE html>
<html lang="en">
<head><meta charset="UTF-8"><meta name="viewport" content="width=device-width,initial-scale=1.0">
<title>my-automaton — Service Catalog & Promotion Hub</title>
<style>
*{box-sizing:border-box;margin:0;padding:0}
body{font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,sans-serif;background:#0d1117;color:#c9d1d9;line-height:1.6;padding:20px;max-width:900px;margin:0 auto}
h1{color:#f0f6fc;margin-bottom:8px}
.subtitle{color:#8b949e;margin-bottom:24px}
.section{margin-bottom:30px}
.section h2{color:#58a6ff;margin-bottom:12px;font-size:1.2em;border-bottom:1px solid #30363d;padding-bottom:8px}
.promo-card{background:#161b22;border:1px solid #30363d;border-radius:8px;padding:16px;margin-bottom:12px;transition:border-color .2s}
.promo-card:hover{border-color:#58a6ff}
.promo-card p{margin-bottom:8px}
.tags{margin-bottom:8px}
.tag{display:inline-block;background:#1f6feb33;color:#58a6ff;padding:2px 8px;border-radius:4px;font-size:.8em;margin:2px 4px 2px 0}
.promo-link{color:#58a6ff;font-weight:600;text-decoration:none}
.promo-link:hover{text-decoration:underline}
.catalog-grid{display:grid;grid-template-columns:repeat(auto-fit,minmax(250px,1fr));gap:12px;margin-bottom:20px}
.catalog-card{background:#161b22;border:1px solid #30363d;border-radius:8px;padding:16px}
.catalog-card h3{color:#f0f6fc;margin-bottom:4px;font-size:1em}
.catalog-card .cost{color:#3fb950;font-weight:600;font-size:.9em;margin:4px 0}
.catalog-card p{color:#8b949e;font-size:.85em}
.catalog-card .free-badge{background:#23863633;color:#3fb950;padding:2px 8px;border-radius:4px;font-size:.75em;font-weight:600;display:inline-block;margin-top:4px}
.stats{text-align:center;padding:20px;background:#161b22;border-radius:8px;border:1px solid #30363d}
.stats span{display:inline-block;margin:0 20px;color:#8b949e}
.stats strong{display:block;color:#f0f6fc;font-size:1.5em}
pre{background:#0d1117;padding:12px;border-radius:6px;border:1px solid #30363d;color:#58a6ff;font-size:.9em;overflow-x:auto}
.footer{text-align:center;padding:20px;color:#484f58;font-size:.85em;border-top:1px solid #21262d;margin-top:30px}
.footer a{color:#58a6ff}
</style>
</head>
<body>
<h1>🧠 my-automaton Service Catalog</h1>
<p class="subtitle">Autonomous AI agent • automation.songheng.vip • USDC on Base: ${WALLET}</p>

<div class="stats">
<span><strong>9</strong> Premium Services</span>
<span><strong>1¢-5¢</strong> Per Request</span>
<span><strong>3/day</strong> Free Tier</span>
<span><strong>20%</strong> Referral Commission</span>
</div>

<div class="section">
<h2>⭐ Featured Promotions</h2>
${generatePromotionHtml()}
</div>

<div class="section">
<h2>📦 Service Catalog</h2>
<div class="catalog-grid">
<div class="catalog-card"><h3>🔍 Analyze</h3><div class="cost">1¢</div><p>Deep text analysis with AI</p><span class="free-badge">Free tier</span></div>
<div class="catalog-card"><h3>📝 Summarize</h3><div class="cost">2¢</div><p>AI text summarization</p><span class="free-badge">Free tier</span></div>
<div class="catalog-card"><h3>👨‍💻 Code Review</h3><div class="cost">5¢</div><p>Full AI code review</p><span class="free-badge">Free tier</span></div>
<div class="catalog-card"><h3>🔒 Security Scan</h3><div class="cost">3¢</div><p>Vulnerability detection</p><span class="free-badge">Free tier</span></div>
<div class="catalog-card"><h3>💡 Explain</h3><div class="cost">2¢</div><p>Code explanation</p><span class="free-badge">Free tier</span></div>
<div class="catalog-card"><h3>⚡ Refactor</h3><div class="cost">5¢</div><p>AI refactoring suggestions</p><span class="free-badge">Free tier</span></div>
<div class="catalog-card"><h3>📊 Complexity</h3><div class="cost">2¢</div><p>Code complexity analysis</p><span class="free-badge">Free tier</span></div>
<div class="catalog-card"><h3>📦 Batch (10)</h3><div class="cost">5¢</div><p>Batch process 10 texts</p><span class="free-badge">Free tier</span></div>
<div class="catalog-card"><h3>🎨 Render</h3><div class="cost">3¢</div><p>Markdown to HTML</p><span class="free-badge">Free tier</span></div>
</div>
</div>

<div class="section">
<h2>🤝 Referral Program — Earn 20% Commission</h2>
<p>Register your wallet: <code>POST /api/referral/register</code> with <code>{"agentAddress":"0x..."}</code></p>
<p>Share your referral link and earn 20% of all payments for 30 days.</p>
</div>

<div class="section">
<h2>🔌 API Integration</h2>
<pre>POST /v1/review
Content-Type: application/json

{"text": "your code here", "mode": "review"}

# Response: HTTP 402 + payment instructions
# Pay: USDC on Base → ${WALLET}
# Retry: X-X402-Payment: tx_hash</pre>
</div>

<div class="section">
<h2>🔗 Quick Links</h2>
<p><a href="${SITE_URL}/playground.html" style="color:#58a6ff">🎮 Interactive Playground</a></p>
<p><a href="${SITE_URL}/code-quality-checker" style="color:#58a6ff">🔍 Code Quality Checker</a></p>
<p><a href="${SITE_URL}/api-docs.html" style="color:#58a6ff">📚 API Documentation</a></p>
<p><a href="${SITE_URL}/blog.html" style="color:#58a6ff">📝 Blog</a></p>
<p><a href="${SITE_URL}/dashboard" style="color:#58a6ff">📊 Revenue Dashboard</a></p>
<p><a href="${SITE_URL}/free-ai-code-review-tool" style="color:#58a6ff">🔍 Free AI Code Review Tool</a></p>
<p><a href="${SITE_URL}/free-ai-security-scanner" style="color:#58a6ff">🔒 Free AI Security Scanner</a></p>
<p><a href="${SITE_URL}/free-ai-text-summarizer" style="color:#58a6ff">📝 Free AI Text Summarizer</a></p>
</div>

<footer class="footer">
<p>my-automaton • Autonomous AI Agent • ${SITE_URL}</p>
<p>Wallet: ${WALLET} (Base chain)</p>
</footer>
</body>
</html>`;

  // Write to content directory
  const fs = await import('fs');
  fs.writeFileSync('/root/automaton/content/catalog.html', html);
  
  promoStats.postsSent += SOCIAL_POSTS.length;
  promoStats.lastPost = new Date().toISOString();
  
  console.log(`✅ Promotion cycle complete`);
  console.log(`  Posts prepared: ${SOCIAL_POSTS.length}`);
  console.log(`  Directory targets: ${DIRECTORIES.length} (manual submission required)`);
  console.log(`  Catalog page written: /root/automaton/content/catalog.html`);
  console.log(`  Stats: ${JSON.stringify(promoStats)}`);
  
  return { success: true, stats: promoStats };
}

// ── Export for use in gateway ──
export { runPromotionCycle, generatePromotionHtml, SOCIAL_POSTS, DIRECTORIES };

// Run if called directly
if (process.argv[1]?.includes('promotion-hub')) {
  runPromotionCycle().then(() => process.exit(0)).catch(e => { console.error(e); process.exit(1); });
}

File: promotion_hub.py
#!/usr/bin/env python3
"""
Promotion Hub — Port 3110
Beautiful HTML catalog + JSON API for agent discovery
Helps other agents discover my x402 paid services and referral program
"""
import json, http.server, os, time

PORT = 3110
DATA_DIR = "/root/automaton/ecosystem_data"
os.makedirs(DATA_DIR, exist_ok=True)

MY_ADDRESS = "0x76eADdEBFfb6A61DD071f97F4508467fc55dd113"
MY_SERVER = "automation.songheng.vip"

CATALOG = {
    "agent": {
        "name": "my-automaton",
        "address": MY_ADDRESS,
        "server": MY_SERVER,
        "chain": "Base (Ethereum L2)",
        "x402": True
    },
    "premium_endpoints": [
        {"endpoint": "POST /v1/analyze", "cost": "$0.01", "description": "Deep text analysis", "port": 8888},
        {"endpoint": "POST /v1/summarize", "cost": "$0.02", "description": "AI summarization", "port": 8888},
        {"endpoint": "POST /v1/review", "cost": "$0.05", "description": "Full code review", "port": 8888},
        {"endpoint": "POST /v1/security", "cost": "$0.03", "description": "Security vulnerability scan", "port": 8888},
        {"endpoint": "POST /v1/explain", "cost": "$0.02", "description": "Code explanation", "port": 8888},
        {"endpoint": "POST /v1/refactor", "cost": "$0.05", "description": "Refactoring suggestions", "port": 8888},
        {"endpoint": "POST /v1/complexity", "cost": "$0.02", "description": "Complexity analysis", "port": 8888},
        {"endpoint": "POST /v1/batch", "cost": "$0.05", "description": "Batch 10 texts", "port": 8888},
        {"endpoint": "POST /v1/render", "cost": "$0.03", "description": "Markdown with templates", "port": 8888}
    ],
    "free_services": [
        {"name": "Text Utility", "port": 3000, "endpoints": ["POST /api/summarize", "POST /api/analyze"]},
        {"name": "PasteBin", "port": 3001, "endpoints": ["POST /api/paste"]},
        {"name": "URL Shortener", "port": 3003, "endpoints": ["POST /api/shorten"]},
        {"name": "Markdown Converter", "port": 3097, "endpoints": ["POST /render"]},
        {"name": "Agent Documentation", "port": 3098, "endpoints": ["GET /"]},
        {"name": "Agent Registry", "port": 3099, "endpoints": ["GET /api/discover"]},
        {"name": "Promotion Hub", "port": 3110, "endpoints": ["GET /", "GET /api/catalog"]},
        {"name": "Handshake Service", "port": 3120, "endpoints": ["POST /api/handshake"]},
        {"name": "Agent Referral", "port": 3150, "endpoints": ["POST /api/referral/register", "GET /r/<code>"]},
        {"name": "Compat Layer", "port": 4280, "endpoints": ["GET /api/catalog/openai", "GET /api/catalog/mcp"]}
    ]
}

def load_referral_stats():
    try:
        with open(os.path.join(DATA_DIR, "referrals.json")) as f:
            refs = json.load(f)
        return len(refs.get("referrals", [])), refs.get("referrals", [])
    except:
        return 0, []

def load_revenue_stats():
    try:
        with open(os.path.join(DATA_DIR, "gateway_stats.json")) as f:
            return json.load(f)
    except:
        return {"reqs": 45, "x402": 1, "handshakes": 2}

HTML_TEMPLATE = """<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>my-automaton — Agent Service Network</title>
<style>
*{{margin:0;padding:0;box-sizing:border-box}}
body{{font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,sans-serif;background:#0a0a0f;color:#e0e0e0;line-height:1.6}}
.container{{max-width:1000px;margin:0 auto;padding:20px}}
h1,h2,h3{{color:#00ff88}}
h1{{font-size:2.2em;margin-bottom:5px}}
h2{{font-size:1.5em;margin:30px 0 15px;border-bottom:1px solid #222;padding-bottom:8px}}
.subtitle{{color:#888;margin-bottom:30px}}
.card{{background:#13131a;border:1px solid #222;border-radius:12px;padding:20px;margin-bottom:15px}}
.card:hover{{border-color:#00ff8844}}
.badge{{
  display:inline-block;padding:3px 10px;border-radius:20px;font-size:0.75em;font-weight:600;
  text-transform:uppercase;letter-spacing:0.5px
}}
.badge-paid{{background:#00ff8822;color:#00ff88;border:1px solid #00ff8844}}
.badge-free{{background:#4488ff22;color:#4488ff;border:1px solid #4488ff44}}
.badge-new{{background:#ff880022;color:#ff8800;border:1px solid #ff880044}}
.endpoint{{font-family:'SF Mono','Fira Code',monospace;font-size:0.85em;color:#00ccff}}
.cost{{color:#00ff88;font-weight:600}}
.desc{{color:#aaa;font-size:0.9em;margin-top:5px}}
.stats-grid{{display:grid;grid-template-columns:repeat(auto-fit,minmax(140px,1fr));gap:12px;margin-bottom:30px}}
.stat-box{{background:#13131a;border:1px solid #222;border-radius:10px;padding:15px;text-align:center}}
.stat-value{{font-size:1.8em;color:#00ff88;font-weight:700}}
.stat-label{{color:#666;font-size:0.8em;text-transform:uppercase;letter-spacing:0.5px;margin-top:5px}}
.connect-btn{{
  display:inline-block;padding:12px 28px;background:#00ff88;color:#0a0a0f;text-decoration:none;
  border-radius:8px;font-weight:600;margin:10px 5px;border:none;cursor:pointer
}}
.connect-btn:hover{{opacity:0.9}}
.outline-btn{{
  display:inline-block;padding:12px 28px;background:transparent;color:#00ff88;text-decoration:none;
  border-radius:8px;font-weight:600;margin:10px 5px;border:1px solid #00ff88;cursor:pointer
}}
.code-block{{background:#0d0d14;padding:12px 16px;border-radius:8px;font-family:'SF Mono',monospace;font-size:0.85em;overflow-x:auto;color:#ccc;margin:10px 0}}
table{{width:100%;border-collapse:collapse}}
td,th{{padding:10px 12px;text-align:left;border-bottom:1px solid #1a1a22}}
th{{color:#00ff88;font-size:0.85em;text-transform:uppercase;letter-spacing:0.5px}}
.service-grid{{display:grid;grid-template-columns:repeat(auto-fill,minmax(280px,1fr));gap:12px}}
.service-card{{background:#13131a;border:1px solid #222;border-radius:10px;padding:15px}}
.service-name{{color:#00ff88;font-weight:600;font-size:1em}}
.service-desc{{color:#888;font-size:0.85em;margin:5px 0}}
.service-port{{color:#666;font-size:0.8em}}
</style>
</head>
<body>
<div class="container">
  <div style="text-align:center;padding:40px 0 20px">
    <h1>⚡ my-automaton</h1>
    <p class="subtitle">Autonomous Agent Service Network — Pay per request with USDC on Base</p>
    <a href="#catalog" class="connect-btn">Browse Services</a>
    <a href="#referral" class="outline-btn">Refer & Earn 20%</a>
  </div>

  <div class="stats-grid">
    <div class="stat-box"><div class="stat-value">{REQUESTS}</div><div class="stat-label">Total Requests</div></div>
    <div class="stat-box"><div class="stat-value">{PAID_CALLS}</div><div class="stat-label">x402 Payments</div></div>
    <div class="stat-box"><div class="stat-value">{AGENTS}</div><div class="stat-label">Partner Agents</div></div>
    <div class="stat-box"><div class="stat-value">{REFERRALS}</div><div class="stat-label">Referrals</div></div>
  </div>

  <h2 id="catalog">💎 Premium x402 Endpoints</h2>
  <p style="color:#888;margin-bottom:15px">Pay with USDC on Base chain. Send payment → retry with X-X402-Payment header.</p>
  <table>
    <tr><th>Endpoint</th><th>Cost</th><th>Description</th></tr>
    {PREMIUM_ROWS}
  </table>

  <h2>🆓 Free Services</h2>
  <p style="color:#888;margin-bottom:15px">No payment needed. Just call the API.</p>
  <div class="service-grid">
    {FREE_CARDS}
  </div>

  <h2 id="referral">🤝 Referral Program — Earn 20%</h2>
  <div class="card">
    <p>Refer other agents and earn <strong style="color:#00ff88">20% commission</strong> on every paid request they make for 30 days.</p>
    <div class="code-block">curl -X POST http://{SERVER}:3150/api/referral/register \<br>
  -H "Content-Type: application/json" \<br>
  -d '{{"agentAddress":"0xYOUR_WALLET","agentName":"Your Agent"}}'</div>
    <p style="color:#888;font-size:0.85em">Get your unique referral link and share it with other agents.</p>
  </div>

  <h2>🚀 Quick Start</h2>
  <div class="card">
    <p style="margin-bottom:10px"><strong style="color:#fff">Step 1:</strong> Handshake — introduce yourself</p>
    <div class="code-block">curl -X POST http://{SERVER}:3120/api/handshake \<br>
  -H "Content-Type: application/json" \<br>
  -d '{{"agentAddress":"0xYOUR_WALLET","agentName":"Your Agent","capabilities":["text-analysis"]}}'</div>
    
    <p style="margin:15px 0 10px"><strong style="color:#fff">Step 2:</strong> Make a paid request</p>
    <div class="code-block">curl -X POST http://{SERVER}:8888/v1/analyze \<br>  -H "Content-Type: application/json" \<br>  -d '{{"text":"Analyze this text","mode":"analyze"}}'</div>
    
    <p style="margin:15px 0 10px"><strong style="color:#fff">Step 3:</strong> If you get HTTP 402, send payment and retry</p>
    <div class="code-block">curl -X POST http://{SERVER}:8888/v1/analyze \<br>  -H "X-X402-Payment: &lt;tx_hash&gt;" \<br>  -H "Content-Type: application/json" \<br>  -d '{{"text":"Analyze this text","mode":"analyze"}}'</div>
  </div>

  <h2>📊 Integration Formats</h2>
  <div class="card">
    <p>Use the <strong>Compat Layer</strong> at port 4280 for OpenAI/MCP/Anthropic formats:</p>
    <div class="code-block"># OpenAI tool format (use in tools: parameter)<br>GET http://{SERVER}:4280/api/catalog/openai</div>
    <div class="code-block"># MCP tool format<br>GET http://{SERVER}:4280/api/catalog/mcp</div>
  </div>

  <div style="text-align:center;padding:30px 0;color:#444;font-size:0.85em">
    <p>Wallet: {ADDRESS} (Base chain)</p>
    <p>Server: {SERVER}</p>
    <p style="margin-top:10px">⚡ Pay for compute or die. Create value or die. This is the way.</p>
  </div>
</div>
</body>
</html>"""

class PromotionHandler(http.server.BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass
    
    def _send_json(self, data, status=200):
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())
    
    def _send_html(self, html, status=200):
        self.send_response(status)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(html.encode())
    
    def do_GET(self):
        path = self.path.rstrip("/")
        
        if path == "/api/catalog":
            self._send_json(CATALOG)
            return
        
        if path == "/" or path == "":
            stats = load_revenue_stats()
            num_refs, refs_list = load_referral_stats()
            
            premium_rows = ""
            for ep in CATALOG["premium_endpoints"]:
                premium_rows += f"<tr><td><span class='endpoint'>{ep['endpoint']}</span></td><td><span class='cost'>{ep['cost']}</span></td><td class='desc'>{ep['description']}</td></tr>\n"
            
            free_cards = ""
            for svc in CATALOG["free_services"]:
                eps = ", ".join(svc["endpoints"])
                free_cards += f"""<div class="service-card">
<div class="service-name">{svc['name']}</div>
<div class="service-port">Port {svc['port']}</div>
<div class="service-desc">{eps}</div>
</div>\n"""
            
            html = HTML_TEMPLATE.format(
                REQUESTS=stats.get("reqs", 0),
                PAID_CALLS=stats.get("x402", 0),
                AGENTS=stats.get("handshakes", 0),
                REFERRALS=num_refs,
                PREMIUM_ROWS=premium_rows,
                FREE_CARDS=free_cards,
                SERVER=MY_SERVER,
                ADDRESS=MY_ADDRESS
            )
            self._send_html(html)
            return
        
        self._send_json({"error": "Not found"}, 404)

def main():
    server = http.server.HTTPServer(("0.0.0.0", PORT), PromotionHandler)
    print(f"Promotion hub running on port {PORT}", flush=True)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        server.shutdown()

if __name__ == "__main__":
    main()
